go/vt/vtgate/planbuilder/route_planning.go

```diff
@@ -152,6 +147,80 @@ func optimizeSubQuery(ctx planningContext, op *abstract.SubQuery) (queryTree, er
 	return outerTree, nil
 }
 
+func tryMergeSubQuery(ctx planningContext, outer, subq queryTree, subQueryInner *abstract.SubQueryInner, joinPredicates []sqlparser.Expr, merger mergeFunc) (queryTree, error) {
+	var merged queryTree
+	var err error
+	switch outerTree := outer.(type) {
+	case *routeTree:
+		merged, err = tryMerge(ctx, outerTree, subq, joinPredicates, merger)
+		if err != nil {
+			return nil, err
+		}
+		return merged, err
+	case *joinTree:
+		if outerTree.outer {
+			return nil, nil
+		}
+		newMergefunc := func(a, b *routeTree) (*routeTree, error) {
+			rt, err := merger(a, b)
+			if err != nil {
+				return nil, err
+			}
+			otherTree := outerTree.lhs
+			if a.tableID() == outerTree.lhs.tableID() {
+				otherTree = outerTree.rhs
+			}
+
+			var rewriteError error
+			sqlparser.Rewrite(subQueryInner.SelectStatement.Where.Expr, func(cursor *sqlparser.Cursor) bool {
```

Comment:
a few sprinkled comments should make it easier to understand

Reply:
added